### PR TITLE
Explainer for Aggregate Debug Reporting

### PR DESCRIPTION
**Pull request by linnan-github**
Related to #705

### CODE REVIEW

File: aggregate_debug_reporting.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+          "key_piece": "0x1", // required
+          "value": 123 // required
+       },
+       {  
+         "types": ["unspecified"],
+          "key_piece": "0x7",
+          "value": 789
+        }
+     ], // default to [], i.e. not opt-in to any debug types 
+     "aggregation_coordinator_origin": "https://publickeyservice.msmt.aws.privacysandboxservices.com" // default to implementation-defined default origin
```
Comment (apasel422) [inline aggregate_debug_reporting.md:67]:
```suggestion
     ], // defaults to [], i.e. doesn't opt-in to any debug types 
     "aggregation_coordinator_origin": "https://publickeyservice.msmt.aws.privacysandboxservices.com" // defaults to implementation-defined default origin
```

File: aggregate_debug_reporting.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+  }
+}
+```
+
+The source registration can pre-allocate `budget` from the total L1 budget
+(L1 = 65 536) for debug reporting, and the remaining will be used for aggregate
```
Comment (apasel422) [inline aggregate_debug_reporting.md:73]:
```suggestion
(L1 = 65536) for debug reporting, and the remaining will be used for aggregate
```

File: aggregate_debug_reporting.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+attribution reporting.
+
+The configuration accepts an optional list field `debug_data` to allow
+developers to specify the aggregation key piece as a hex-string and the
+corresponding histogram value for debug types they opt-in to receiving debug
+reports. When "unspecified" value is set, any non-explicitly specified debug
```
Comment (apasel422) [inline aggregate_debug_reporting.md:79]:
```suggestion
corresponding histogram value for debug types they opt-in to receiving.
When the `unspecified` value is set, any non-explicitly specified debug
```

File: aggregate_debug_reporting.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+types will be reported with the corresponding key piece and value, otherwise
```
Comment (apasel422) [inline aggregate_debug_reporting.md:80]:
```suggestion
types will be reported with the corresponding key piece and value; otherwise
```

File: aggregate_debug_reporting.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+they will not be reported. The list of supported debug types will be documented
+in the [specification](https://wicg.github.io/attribution-reporting-api/).
+
+The configuration also accepts an optional string field
+`aggregation_coordinator_origin` to allow developers to specify the deployment
+option for the aggregation service supported by the browser. This is default to
```
Comment (apasel422) [inline aggregate_debug_reporting.md:86]:
```suggestion
option for the aggregation service supported by the browser. This defaults to
```

File: aggregate_debug_reporting.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+the implementation-defined default origin if it’s not present.
```
Comment (apasel422) [inline aggregate_debug_reporting.md:87]:
```suggestion
the implementation-defined default origin if omitted.
```

File: aggregate_debug_reporting.md [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+The final histogram bucket key is defined as a combination (bitwise OR) of the
+source-side key piece if available,  the trigger-side key piece if available,
```
Comment (apasel422) [inline aggregate_debug_reporting.md:90]:
```suggestion
source-side key piece if available, the trigger-side key piece if available,
```

File: aggregate_debug_reporting.md [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+and [trigger registration](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#triggering-attribution)
+will accept an optional dictionary field:
+
+```jsonc
+{
+  ..., // existing fields
+
+  "aggregate_debug_reporting": {
+     "budget": 1024, // required, source registration only
+     "key_piece": "0x120", // required, source or trigger side key piece
```
Comment (apasel422) [inline aggregate_debug_reporting.md:54]:
```suggestion
     "key_piece": "0x120", // required, source- or trigger-side key piece
```

File: aggregate_debug_reporting.md [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+      "key_id": "[string identifying public key used to encrypt payload]",
+    },
+  ],
+
+  // The deployment option for the aggregation service.
+  "aggregation_coordinator_origin": "https://publickeyservice.msmt.aws.privacysandboxservices.com"
+}
+```
+
+Note: If the source registration specifies a list of destinations, the smallest
```
Comment (apasel422) [inline aggregate_debug_reporting.md:131]:
```suggestion
Note: If the source registration specifies a list of destinations, the first
```

File: aggregate_debug_reporting.md [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+one in lexicographical order will be used for debug reporting.
+
+#### Encrypted payload
+
+The `payload` should be a [CBOR](https://cbor.io) map encrypted via
+[HPKE](https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/) and then base64-
+encoded. The map will have the following structure:
+
+```
+// CBOR
```
Comment (apasel422) [inline aggregate_debug_reporting.md:141]:
```suggestion
```cbor
```

File: aggregate_debug_reporting.md [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+histogram contributions per source. The budget for debug reporting is
+pre-allocated at source registration time, and exceeding the budget will cause
+future contributions to silently drop.
+
+The aggregation service will add noise to the summary reports to achieve
+differential privacy, e.g. by using Laplace noise scaled based on the L1
+sensitivity and a desired privacy parameter epsilon. The effective epsilon
+needs to be scaled based on the budget allocated for debug reporting.
+
+Note: The L1 budget (65 536) is shared between the aggregatable attribution
```
Comment (apasel422) [inline aggregate_debug_reporting.md:173]:
```suggestion
Note: The L1 budget (65536) is shared between the aggregatable attribution
```

File: aggregate_debug_reporting.md [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+To limit the amount of user-identity leakage, the browser should throttle the 
+amount of total information sent through this API in a given time period for a
+user. The browser could bound the contributions that any
+<top-level site, reporting site> tuple can make to the histogram within a
+rolling time window. If this threshold is hit, the browser will stop scheduling
+aggregatable debug reports for the rest of the time period for attribution
+registrations matching that tuple.
+
+Strawman: L1 bound of 65 536 per <top-level site, reporting site> per day.
```
Comment (apasel422) [inline aggregate_debug_reporting.md:205]:
```suggestion
Strawman: L1 bound of 65536 per <top-level site, reporting site> per day.
```

File: aggregate_debug_reporting.md [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+
+Additionally, to mitigate the attack in which multiple reporting sites collude
+with each other to infer user data from the debug reports, the browser can also
+bound the contributions that any top-level site can make to the histogram
+within a rolling time window.
+
+Strawman: L1 bound of 2<sup>20</sup> = 1 048 576 per top-level site per day.
```
Comment (apasel422) [inline aggregate_debug_reporting.md:212]:
```suggestion
Strawman: L1 bound of 2<sup>20</sup> = 1048576 per top-level site per day.
```

File: aggregate_debug_reporting.md [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Aggregate Debug Reporting](#aggregate-debug-reporting)
+  - [Introduction](#introduction)
+  - [API changes](#api-changes)
+    - [Opting-in aggregate debug reporting](#opting-in-aggregate-debug-reporting)
+    - [Aggregatable debug  reports](#aggregatable-debug--reports)
```
Comment (apasel422) [inline aggregate_debug_reporting.md:11]:
```suggestion
    - [Aggregatable debug reports](#aggregatable-debug--reports)
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: aggregate_debug_reporting.md [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+     "budget": 1024, // required, source registration only
+     "key_piece": "0x120", // required, source- or trigger-side key piece
+     "debug_data": [    
+        {  
+          "types": ["source-destination-limit",  "source-destination-rate-limit"], // required to be non-empty
+          "key_piece": "0x1", // required
+          "value": 123 // required
+       },
+       {  
+         "types": ["unspecified"],
```
Comment (apasel422) [inline aggregate_debug_reporting.md:62]: I find the term `unspecified` to be a bit vague. I think it could be misinterpreted as "do anything". I wonder if a more conventional wildcard makes more sense, e.g. `*`, which is also a bit more concise and will pretty clearly never conflict with an actual debug report type.

File: aggregate_debug_reporting.md [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+The [source registration](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#registering-attribution-sources)
+and [trigger registration](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#triggering-attribution)
+will accept an optional dictionary field:
+
+```jsonc
+{
+  ..., // existing fields
+
+  "aggregatable_debug_reporting": {
+     "budget": 1024, // required, source registration only
```
Comment (apasel422) [inline aggregate_debug_reporting.md:53]: The indentation here and throughout this jsonc snippet is off: It seems to be using 3 spaces sometimes instead of 2 throughout.

File: aggregate_debug_reporting.md [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+     "key_piece": "0x120", // required, source- or trigger-side key piece
+     "debug_data": [    
+        {  
+          "types": ["source-destination-limit",  "source-destination-rate-limit"], // required to be non-empty
```
Comment (apasel422) [inline aggregate_debug_reporting.md:57]:
```suggestion
          "types": ["source-destination-limit", "source-destination-rate-limit"], // required to be non-empty
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: aggregate_debug_reporting.md [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    "debug_data": [    
+      {  
+        "types": ["source-destination-limit", "source-destination-rate-limit"], // required to be non-empty
+        "key_piece": "0x1", // required
+        "value": 123 // required
+      },
+      {  
+        "types": ["unspecified"],
+         "key_piece": "0x7",
+         "value": 789
```
Comment (apasel422) [inline aggregate_debug_reporting.md:64]:
```suggestion
        "key_piece": "0x7",
        "value": 789
```

File: aggregate_debug_reporting.md [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+(L1 = 65536) for debug reporting, and the remaining will be used for aggregate
+attribution reporting.
+
+The configuration accepts an optional list field `debug_data` to allow
+developers to specify the aggregation key piece as a hex-string and the
+corresponding histogram value for debug types they opt-in to receiving.
+When the `unspecified` value is set, any non-explicitly specified debug
+types will be reported with the corresponding key piece and value; otherwise
+they will not be reported. The list of supported debug types will be documented
+in the [specification](https://wicg.github.io/attribution-reporting-api/).
```
Comment (apasel422) [inline aggregate_debug_reporting.md:82]:
```suggestion
in the [specification](https://wicg.github.io/attribution-reporting-api/#attribution-debug-data).
```

File: aggregate_debug_reporting.md [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+The [source registration](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#registering-attribution-sources)
+and [trigger registration](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md#triggering-attribution)
+will accept an optional dictionary field:
+
+```jsonc
+{
+  ..., // existing fields
+
+  "aggregatable_debug_reporting": {
+    "budget": 1024, // required, source registration only
```
Comment (apasel422) [inline aggregate_debug_reporting.md:53]: The comment here may be misinterpreted as "required for sources, optional for triggers". Maybe rephrase as `required for sources, irrelevant for triggers`?

File: aggregate_debug_reporting.md [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+    "key_piece": "0x120", // required, source- or trigger-side key piece
+    "debug_data": [    
+      {  
+        "types": ["source-destination-limit", "source-destination-rate-limit"], // required to be non-empty
```
Comment (apasel422) [inline aggregate_debug_reporting.md:57]:
```suggestion
        "types": ["source-destination-limit", "source-destination-rate-limit"], // required to be present and non-empty
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Review by akashnadan — APPROVED** [review]

**Review by apasel422 — APPROVED** [review]
